### PR TITLE
DrivAerML: Grouped Query Attention (GQA) — fewer KV heads for throughput

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The MoE experiment (#3263) demonstrated that **throughput is the primary constraint** on DM — any modification that slows per-epoch time below baseline gets epoch-starved and loses. The flip side: any modification that IMPROVES throughput gets more epochs within the timeout and may find deeper optima.

**Grouped Query Attention (GQA)** (Ainslie et al., 2023; used in LLaMA 2/3, Mistral, Gemma) reduces the number of Key-Value heads while keeping all Query heads. With 8 Q heads and 2 KV groups, each KV head is shared across 4 Q heads. This reduces the KV projection parameters by 4x and the KV computation by 4x, while maintaining the full Q-head expressivity for attention pattern computation.

For DM with 50k surface point tokens at bs=1:
- Standard MHA: 8 Q heads × 64d, 8 K heads × 64d, 8 V heads × 64d → 3 × 512 → 512 projections
- GQA (2 KV groups): 8 Q heads × 64d, 2 K heads × 64d, 2 V heads × 64d → Q: 512→512, K: 512→128, V: 512→128

This saves ~25% of the attention parameter budget and reduces memory bandwidth for the KV cache. At 50k tokens, the KV computation is a significant fraction of total FLOPS. If GQA provides even 10-15% throughput improvement, that translates to ~50-75 more epochs within the timeout — potentially enough to push past 3.833%.

**Key insight from the head count sweep:** 8H is the optimal head count. GQA tests whether the VALUE of all 8 heads matters (it might) or whether 8 QUERY patterns are sufficient even with shared KV projections (the GQA hypothesis).

### Implementation

1. **Add CLI flag:**
   ```python
   parser.add_argument('--num-kv-heads', type=int, default=0,
                       help='Number of KV heads for GQA (0=same as model-heads, standard MHA)')
   ```

2. **Modify the self-attention module** to support GQA:
   ```python
   num_q_heads = args.model_heads  # 8
   num_kv_heads = args.num_kv_heads if args.num_kv_heads > 0 else num_q_heads
   head_dim = hidden_dim // num_q_heads  # 64
   
   # Q projection: full heads
   self.q_proj = nn.Linear(hidden_dim, num_q_heads * head_dim)
   # K, V projections: fewer heads
   self.k_proj = nn.Linear(hidden_dim, num_kv_heads * head_dim)
   self.v_proj = nn.Linear(hidden_dim, num_kv_heads * head_dim)
   self.o_proj = nn.Linear(num_q_heads * head_dim, hidden_dim)
   ```

3. **In the attention forward pass**, repeat KV heads to match Q heads:
   ```python
   # After projecting Q, K, V:
   # Q: [B, N, num_q_heads, head_dim]
   # K: [B, N, num_kv_heads, head_dim]
   # V: [B, N, num_kv_heads, head_dim]
   
   if num_kv_heads < num_q_heads:
       repeat_factor = num_q_heads // num_kv_heads
       K = K.repeat_interleave(repeat_factor, dim=2)  # [B, N, num_q_heads, head_dim]
       V = V.repeat_interleave(repeat_factor, dim=2)
   
   # Then compute attention normally with Q, K, V all having num_q_heads
   ```

   **Alternative (more efficient):** Use `torch.nn.functional.scaled_dot_product_attention` with GQA-aware reshaping, which avoids the explicit repeat.

### Trials

**Smoke test first (3 epochs)** — verify shapes, attention output matches expected dims, loss finite. Also measure per-epoch time vs baseline.

**Trial 1 — 2 KV heads (4:1 Q/KV ratio):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --num-kv-heads 2 \
  --wandb_group dm-gqa
```

**Trial 2 — 4 KV heads (2:1 Q/KV ratio, milder):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --num-kv-heads 4 \
  --wandb_group dm-gqa
```

**Critical:** Report per-epoch time (min/epoch) for both trials alongside the standard metrics. The throughput improvement IS the hypothesis — if GQA doesn't speed things up measurably, the experiment is moot.

Report `val_primary/surface_rel_l2_pct` for both trials. Target: beat **3.833%**.

## Baseline

Current best DrivAerML metrics (PR #3072):

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_rel_l2_pct | **3.833%** | 511 |
| test_primary/surface_rel_l2_pct | **4.685%** | 511 |

- W&B run: `ncl1dh88`
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, gc=0.5, EMA=0.9995, no WD, Fourier

Reproduce:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT gojo:

## Results

### Summary

| Config | Params | Avg epoch time | Best val `surface_rel_l2_pct` | Best epoch | Diverged at | W&B run |
|--------|--------|---------------|-------------------------------|------------|-------------|---------|
| **Baseline (8H MHA)** | 12,776,620 | ~0.70 min | **3.833%** | — | — | — |
| **Trial 1: 2 KV heads** | 11,988,652 (−6.2%) | ~0.71 min | 6.884% | 114 | ~epoch 120 | `rq2tz9hh` |
| **Trial 2: 4 KV heads** | 12,251,308 (−4.1%) | ~0.73 min | 12.882% | 57 | ~epoch 57 | `wbygalr3` |

### Throughput hypothesis: **falsified**

GQA is **throughput-neutral** on this architecture. Per-epoch time:
- Baseline MHA: ~0.70 min/epoch
- GQA-2KV: ~0.71 min/epoch (±0%)
- GQA-4KV: ~0.73 min/epoch (±0%)

The 6% parameter reduction does not translate to wall-clock savings. The bottleneck is the 50k-token einsums for slice routing (`in_project_x`, `in_project_slice`, slice aggregation) and the MLP layers, not the KV projections on 96 slice tokens.

### Quality: **significant regression**

Neither trial approaches the 3.833% baseline:
- 2KV best: **6.884%** (80% worse than baseline)
- 4KV best: **12.882%** (236% worse than baseline)

### Training instability: both configs diverge

Both GQA configs eventually hit catastrophic gradient norm explosions:

**Trial 2 (4KV):** Diverged at epoch 57-58. Grad norm spiked 8→23→52, then settled into permanent oscillation (grad_norm 30-170). EMA preserved best at 12.88%.

**Trial 1 (2KV):** Stable through 114 epochs with smooth descent (best 6.88%). At the 4th cosine cycle restart (epoch 120), grad_norm began escalating: 3→6→8→10→20→46→147→180. The online model diverged catastrophically. EMA preserved best at 6.88%.

The pattern suggests GQA's averaged group slice_weights for KV routing introduces fragile training dynamics in the Transolver's slice-based attention. Sharing KV projections across Q head groups means each Q head group's routing must be consistent enough that the averaged KV routing still produces useful representations — this may create coupled optimization dynamics that become unstable at higher learning rates.

### Commands used

```bash
# Trial 1: 2 KV heads
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --num-kv-heads 2 \
  --agent gojo --wandb-name "gojo/dm-gqa-2kv" --wandb-group dm-gqa

# Trial 2: 4 KV heads
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --num-kv-heads 4 \
  --agent gojo --wandb-name "gojo/dm-gqa-4kv" --wandb-group dm-gqa
```

- Peak VRAM: ~5-6 GB per GPU (estimated, not explicitly logged)
- W&B runs: [`rq2tz9hh`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/rq2tz9hh), [`wbygalr3`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/wbygalr3)
- Processes killed at epoch 133 (2KV) and epoch 129 (4KV) after both fully diverged

### What happened

The throughput hypothesis is **dead**. GQA saves parameters but not wall-clock time on this architecture because the computational bottleneck is the Transolver's unique slice routing mechanism (50k-token einsums), not the QKV projections which operate on only 96 slice tokens.

Worse, GQA introduces training instability. The Transolver uses learned slice weights to route 50k tokens into 96 slice tokens — each head computes its own routing. With GQA, KV heads share averaged routing weights across Q head groups, creating coupled optimization dynamics that eventually diverge. The 4KV config (2:1 ratio, milder) diverged earlier than 2KV (4:1 ratio), suggesting the shared routing is the root cause — with more KV heads, each shared group is smaller but there are more of them, leading to more frequent instability.

### Suggested follow-ups

1. **This idea is dead for DrivAerML** — the throughput hypothesis failed and quality regressed significantly.
2. If throughput improvement is still the goal, the bottleneck is the slice routing einsums, not the attention projections. Techniques like sparse slice routing or fewer slices might be more productive.
3. The divergence pattern (stable for many epochs, then catastrophic) might warrant investigation of gradient norm monitoring with early stopping as a general training improvement, independent of GQA.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3273 because ADVISOR: Closing — dead end. 2KV: 12.882% (+236%), 4KV: 6.884% (+80%), both crashed. GQA is throughput-neutral here because the real bottleneck is the Transolver's 50k-token slice-routing einsums, not the QKV projections operating on only 96 slice tokens. GQA also destabilizes per-head slice routing — KV sharing forces coupled routing dynamics that explode at cosine restart peaks. Excellent diagnostic work identifying the true throughput bottleneck. Key insight preserved: future throughput experiments should target slice routing, not attention projections..
